client/app/bundles/course/assessment/reducers/monitoring.ts

```diff
@@ -0,0 +1,88 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+import {
+  HeartbeatDetail,
+  MonitoringMonitorData,
+  Snapshot,
+  Snapshots,
+} from 'types/channels/liveMonitoring';
+
+import { Activity, MonitoringState } from '../pages/AssessmentMonitoring/types';
+
+const initialState: MonitoringState = {
+  snapshots: {},
+  history: [],
+  connected: false,
+  monitor: { maxIntervalMs: 0, offsetMs: 0, hasSEBHash: false },
+};
+
+export const monitoringSlice = createSlice({
+  name: 'monitoring',
+  initialState,
+  reducers: {
+    initialize: (
+      state,
+      action: PayloadAction<{
+        monitor: MonitoringMonitorData;
+        snapshots: Snapshots;
+      }>,
+    ) => {
+      state.monitor = action.payload.monitor;
+      state.snapshots = action.payload.snapshots;
+    },
+    refresh: (
+      state,
+      action: PayloadAction<{ userId: number; data: Partial<Snapshot> }>,
+    ) => {
+      const { userId, data } = action.payload;
+      const snapshot = state.snapshots[userId];
+
+      state.snapshots[userId] = { ...snapshot, ...data };
+    },
+    pushHistory: (state, action: PayloadAction<Activity>) => {
+      state.history.push(action.payload);
+    },
+    selectSnapshot: (state, action: PayloadAction<number>) => {
+      const selectedUserId = action.payload;
+      state.selectedUserId = selectedUserId;
+      state.snapshots[selectedUserId].selected = true;
+    },
+    deselectSnapshot: (state) => {
+      const userId = state.selectedUserId;
+      if (!userId) return;
+
+      state.selectedUserId = undefined;
+      delete state.snapshots[userId].selected;
+    },
+    setConnected: (state, action: PayloadAction<boolean>) => {
+      state.connected = action.payload;
+    },
+    terminate: (state, action: PayloadAction<number>) => {
+      const userId = action.payload;
+      state.snapshots[userId].status = 'stopped';
+    },
+    supplySelectedSnapshot: (
+      state,
+      action: PayloadAction<HeartbeatDetail[]>,
+    ) => {
+      const { selectedUserId } = state;
```

Comment:
Technically, while the next `supplySelected` will add the `heartbeats` to the wrong `userId`'s snapshot, it doesn't really matter because:
- since you've clicked the other user, there's going to be another `getRecentHeartbeats` for the new user, hence another `supplySelected` for _the correct user_, and
- even if the old user's `recentHeartbeats` is incorrect, if one were to click again, it will fetch the correct heartbeats again.

But sure, I'll have the server add the `userId` and `supplySelectedSnapshot` reject if `userId !== selectedUserId`.